vowpalwabbit/config/options_cli.cc

```diff
@@ -0,0 +1,486 @@
+    /*in-out*/ std::queue<VW::string_view>& command_line,
+    /*out*/ std::unordered_map<VW::string_view, std::vector<VW::string_view>>& result)
+{
+  auto current_token = command_line.front();
+  auto current_opt = current_token.substr(2);
+  const auto equal_sign_pos = current_opt.find('=');
+  std::vector<VW::string_view> current_tokens;
+  if (equal_sign_pos != std::string::npos)
+  {
+    auto opt_value = current_opt.substr(equal_sign_pos + 1);
+    current_opt = current_opt.substr(0, equal_sign_pos);
+    if (opt_value.empty()) { THROW("Argument for " << current_opt << " should follow =") }
+    current_tokens.push_back(opt_value);
+  }
+
+  if (current_opt.empty()) { THROW("Expected long option name after -- but found nothing") }
+
+  // Consume this token.
+  command_line.pop();
+
+  // See if there is a known option for this name
+  auto known_opt = known_options.find(std::string{current_opt});
+  if (known_opt == known_options.end())
+  {
+    // If there is no known option we are going to just treat it as an unknown option.
+    const auto result_name = "__POSITIONAL__";
+    result[result_name].push_back(current_token);
+    return;
+  }
+
+  const auto& option_spec = *known_opt->second;
+  consume_tokens(option_spec, command_line, current_tokens);
+
+  // Will create vector if not exists - this is intentional.
+  auto& result_tokens = result[option_spec.m_name];
+  for (auto& token : current_tokens) { result_tokens.push_back(token); }
+}
+
+void consume_short_option(const std::map<char, std::shared_ptr<base_option>>& known_short_options,
+    /*in-out*/ std::queue<VW::string_view>& command_line,
+    /*out*/ std::unordered_map<VW::string_view, std::vector<VW::string_view>>& result)
+{
+  auto current_token = command_line.front();
+
+  // Check if just dangling -
+  if (current_token.size() == 1)
+  {
+    const auto result_name = "__POSITIONAL__";
+    result[result_name].push_back(current_token);
+    command_line.pop();
+    return;
+  }
+
+  auto current_opt = current_token[1];
+
+  std::vector<VW::string_view> current_tokens;
+  if (current_token.size() > 2)
+  {
+    auto opt_value = current_token.substr(2);
+    current_tokens.push_back(opt_value);
+  }
+
+  // Consume this token.
+  command_line.pop();
+
+  // See if there is a known option for this name
+  auto known_opt = known_short_options.find(current_opt);
+  if (known_opt == known_short_options.end())
+  {
+    // If there is no known option we are going to just treat it as an unknown option.
+    result["__POSITIONAL__"].push_back(current_token);
+    return;
+  }
+
+  const auto& option_spec = *known_opt->second;
+  consume_tokens(option_spec, command_line, current_tokens);
+
+  // Will create vector if not exists - this is intentional.
+  auto& result_tokens = result[option_spec.m_name];
+  for (auto& token : current_tokens) { result_tokens.push_back(token); }
+}
+
+namespace VW
+{
+namespace config
+{
+namespace details

```

Comment:
`VW::config::details` is a bit generic? maybe `token_parser` or `token_handler` or something else for a namespace?

Reply:
`details` is a standard namespace for implementation details that the end user should not use

Reply:
that's good to know :)